### DISCUSSION TITLE
Supporting additional object_store traits for hyperscaler

### DISCUSSION THREAD

**Discussion by thhapke** [delta-io/delta-rs#2514] [Q&A] [answered]
I have implemented an additional object_store HDFS for using datafusion. 

After I learnt that deltalake version 0.17..0 is currently working only with datafusion 35.0.0, I am running now into the following error:
```
error[E0562]: `impl Trait` only allowed in function and inherent method return types, not in trait method return types
   --> /Users/me/.cargo/registry/src/index.crates.io-6f17d22bba15001f/deltalake-core-0.17.3/src/delta_datafusion/mod.rs:240:73
    |
240 |     fn files_matching_predicate(&self, filters: &[Expr]) -> DeltaResult<impl Iterator<Item = Add>>;
    |                                                                         ^^^^^^^^^^^^^^^^^^^^^^^^^
    |
    = note: see issue #91611 <https://github.com/rust-lang/rust/issues/91611> for more information

error[E0562]: `impl Trait` only allowed in function and inherent method return types, not in `impl` method return types
   --> /Users/me/.cargo/registry/src/index.crates.io-6f17d22bba15001f/deltalake-core-0.17.3/src/delta_datafusion/mod.rs:244:73
    |
244 |     fn files_matching_predicate(&self, filters: &[Expr]) -> DeltaResult<impl Iterator<Item = Add>> {
    |                                                                         ^^^^^^^^^^^^^^^^^^^^^^^^^
    |
    = note: see issue #91611 <https://github.com/rust-lang/rust/issues/91611> for more information
```
Is this caused that the object store is not one of the supported S3, gcp, .. etc?

**Comment by roeap** [comment, marked as answer]
I think you may need to check the version of Rust you are using. IIRC for the `impl Trait` syntax to be allowed in a traits return type you must at least be on [`1.75`](https://blog.rust-lang.org/2023/12/28/Rust-1.75.0.html#async-fn-and-return-position-impl-trait-in-traits) and we use that for internal traits. Which of course implies that our rust version bounds are not correct, since this should also be reflected in the MSRV within our cargo.toml files.
